content/posts/2022-08-14-libc++-format.md

```diff
@@ -0,0 +1,273 @@
+emojis.  (This "combined multiple characters" is known as
+[extended grapheme clusters](https://www.unicode.org/reports/tr29/) in
+Unicode.) The library has implemented these rules so it will count both forms
+of Á as using one column in the output.
+
+Another issue with text formatting is that not all every "character" has the
+same column width. Based on the "character" the column width is estimated to be
+one or two columns. 
+
+Below an example taken from the [paper](https://wg21.link/p1868r2) introducing
+this feature.
+
+```cpp
+struct input {
+  const char* text;
+  const char* info;
+};
+
+int main() {
+  input inputs[] = {
+    {"Text", "Description"},
+    {"-----",
+     "------------------------------------------------------------------------"
+     "--------------"},
+    {"\x41", "U+0041 { LATIN CAPITAL LETTER A }"},
+    {"\xC3\x81", "U+00C1 { LATIN CAPITAL LETTER A WITH ACUTE }"},
+    {"\x41\xCC\x81",
+     "U+0041 U+0301 { LATIN CAPITAL LETTER A } { COMBINING ACUTE ACCENT }"},
+    {"\xc4\xb2", "U+0132 { LATIN CAPITAL LIGATURE IJ }"}, // Ĳ
+    {"\xce\x94", "U+0394 { GREEK CAPITAL LETTER DELTA }"}, // Δ
+    {"\xd0\xa9", "U+0429 { CYRILLIC CAPITAL LETTER SHCHA }"}, // Щ
+    {"\xd7\x90", "U+05D0 { HEBREW LETTER ALEF }"}, // א
+    {"\xd8\xb4", "U+0634 { ARABIC LETTER SHEEN }"}, // ش
+    {"\xe3\x80\x89", "U+3009 { RIGHT-POINTING ANGLE BRACKET }"}, // 〉
+    {"\xe7\x95\x8c", "U+754C { CJK Unified Ideograph-754C }"}, // 界
+    {"\xf0\x9f\xa6\x84", "U+1F921 { UNICORN FACE }"}, // 🦄
+    {"\xf0\x9f\x91\xa8\xe2\x80\x8d\xf0\x9f\x91\xa9\xe2\x80\x8d"
+     "\xf0\x9f\x91\xa7\xe2\x80\x8d\xf0\x9f\x91\xa6",
+     "U+1F468 U+200D U+1F469 U+200D U+1F467 U+200D U+1F466 "
+     "{ Family: Man, Woman, Girl, Boy } "} // 👨‍👩‍👧‍👦
+  };
+
+  for (auto input: inputs) {
+    std::cout << std::format("{:>5} | {}\n", input.text, input.info);
+  }
+}
+```
+(Note the column width is intended to look good on a terminal. The author has
+observed differences in quality of the output depending on the browser used.)
+```
+ Text | Description
+----- | --------------------------------------------------------------------------------------
+    A | U+0041 { LATIN CAPITAL LETTER A }
+    Á | U+00C1 { LATIN CAPITAL LETTER A WITH ACUTE }
+    Á | U+0041 U+0301 { LATIN CAPITAL LETTER A } { COMBINING ACUTE ACCENT }
+    Ĳ | U+0132 { LATIN CAPITAL LIGATURE IJ }
+    Δ | U+0394 { GREEK CAPITAL LETTER DELTA }
+    Щ | U+0429 { CYRILLIC CAPITAL LETTER SHCHA }
+    א | U+05D0 { HEBREW LETTER ALEF }
+    ش | U+0634 { ARABIC LETTER SHEEN }
+   〉 | U+3009 { RIGHT-POINTING ANGLE BRACKET }
+   界 | U+754C { CJK Unified Ideograph-754C }
+   🦄 | U+1F921 { UNICORN FACE }
+   👨‍👩‍👧‍👦 | U+1F468 U+200D U+1F469 U+200D U+1F467 U+200D U+1F466 { Family: Man, Woman, Girl, Boy } 
+```
+
+A huge benefit of ``std::format`` over ``printf`` is its type safety.
+When attempting to format a string as a number, it will result in a compilation
+failure.
```

Comment:
consteval is introduced in C++20. I'm not convinced that detail is really required. I keep it since it might get people interested in consteval, which is a very nice language feature.